HTTP CONNECT proxy.

arut opened the pull request:
[HTTP CONNECT](https://www.rfc-editor.org/rfc/rfc9110#CONNECT) method is now supported in HTTP/1 connections. A new module ngx_http_tunnel module is added which establishes a tunnel to a backend.  It supports the newly added HTTP CONNECT method and can be used to set up an HTTP CONNECT proxy.

Documentation: https://nginx.org/en/docs/http/ngx_http_tunnel_module.html

Forward proxy with host/port validation:
```
error_log logs/error.log debug;                                                  
                                                                      
events { }                                                                                
                                                                                 
http {                                                                           
    server {
        listen 8000;

        resolver dns.example.com;

        map $port $tun_port {
            80             1;
            443            1;
        }

        map $host $tun_host {
            hostnames;

            example.com    1;
            *.example.org  1;
        }

        map $tun_port$tun_host $tun {
            11             $host:$port;
        }

        location / {
            tunnel_pass $tun;
        }
    }
}
```

```bash
$ curl -x 127.0.0.1:8000 https://example.com
```

A simple HTTPS proxy with authentication:
```
error_log logs/error.log debug;                                                  
                                                                      
events { }                                                                                
                                                                                 
http {                                                                                                                                                
    log_format  main  '$remote_addr - $request_method - $remote_user '           
                      '[$time_local] "$request" '                                
                      '$status $body_bytes_sent "$http_referer" '                
                      '"$http_user_agent"';                                      
                                                                                 
    access_log  logs/access.log  main;                                           
                                                                                 
    server {                                                                     
        listen 8800;                                                             
                                                                                 
        resolver 1.1.1.1;                                                        
                                                                                 
        auth_basic "closed site";                                                
        auth_basic_user_file htpasswd;                                           
                                                                                 
        tunnel_pass;                                                             
    }                                                                            
}
```

```bash
$ curl --user myuser --pass mypass -x 127.0.0.1:8800 https://example.com
```

jo-carter commented:
@arut should [limit_except](https://nginx.org/en/docs/http/ngx_http_core_module.html#limit_except) be adjusted to include CONNECT, as part of this changeset?

Also was performing virtual server lookup using the destination's host/ip component intended?

arut commented:
> @arut should [limit_except](https://nginx.org/en/docs/http/ngx_http_core_module.html#limit_except) be adjusted to include CONNECT, as part of this changeset?

Yes, thanks. Added.

> Also was performing virtual server lookup using the destination's host/ip component intended?

I thought about disabling the lookup for CONNECT completely. However it's just more code, and it seems like there's actually a use case for that, which is to restrict proxying to a predefined set of hosts. And it's not host+ip, it's just host, as usual.

An example which restricts proxying to just 2 servers (port 80 hardcoded). More server-specific settings can be added.
```
    server {                                                                     
        listen 8000;
        return 502;
    }

    server {                                                                     
        listen 8000;                                                                                                                       
        server_name nginx.com;                                                      
                                                                                 
        location / {                                                             
            tunnel_pass nginx.com:80;                                               
        }                                                                        
    } 
    server {                                                                     
        listen 8000;                                                                                                                                     
        server_name f5.com;                                                      
                                                                                 
        location / {                                                             
            tunnel_pass f5.com:80;                                               
        }                                                                        
    } 
```

arut committed:
HTTP tunnel module

The module handles CONNECT requests and establishes a tunnel to a
backend.

Example config:

http {

    map $request_port $allow_port {
        80             1;
        443            1;
    }

    map $host $allow_host {
        hostnames;

        example.com    1;
        *.example.org  1;
    }

    server {
        listen 8000;

        resolver dns.example.com;

        if ($allow_port != 1) {
            return 403;
        }

        if ($allow_host != 1) {
            return 403;
        }

        tunnel_pass;
    }
}

Request:

    $ curl -x 127.0.0.1:8000 https://example.com

arut committed:
Proxy authentication for CONNECT requests

Notably, ngx_http_auth_basic_module uses Proxy-Authorization input
header, Proxy-Authenticate output header and HTTP code 407 instead
of Authorization, WWW-Authenticate and 401 respectively.

sindhushiv commented:
Closes https://github.com/nginx/nginx/issues/1140 on GitHub.

dplotnikov-f5 commented on src/http/ngx_http_core_module.c:
We should support `satisfy any` for `NGX_HTTP_PROXY_AUTH_REQUIRED` as well.

arut replied:
Yes. Updated.

dplotnikov-f5 commented on src/http/ngx_http_core_module.c:
`NGX_HTTP_PROXY_AUTH_REQUIRED` should support `auth_delay` also (here and in `ngx_http_core_post_access_phase()`).

arut replied:
Yes, Updated.

ZihaoFU245 commented:
@arut Hello, I have done some connect header parsing work on #1277 . Any plans for letting this module also work with h2 and h3?

arut committed:
Support 407 code in "satisfy any" and "auth_delay"

Notably, "auth_delay" now delays the response for both 401 and 407.
Also, in the "satisfy any" mode, the next access/auth attempt is made
for 401, 403 and 407.

pluknet commented on src/http/ngx_http_upstream.c:
```diff
@@ -3315,6 +3327,11 @@ ngx_http_upstream_send_response(ngx_http_request_t *r, ngx_http_upstream_t *u)
+            ngx_http_upstream_finalize_request(r, u, 0);
+            return;
+        }
+
```
Is it really needed now?

The only consumer of ignore_input is the tunnel module, which sets u->upgrade unconditionally.
So we never get there.

arut replied:
Thanks. Removed.

pluknet reviewed (commented):
It is not immediately obvious from the first commit log
that the tunnel directive optionally accepts the target host argument,
with the default to $host:$request_port".

pluknet commented on src/http/ngx_http_request.h:
```diff
@@ -620,6 +620,8 @@ typedef struct {
 
 #define ngx_http_ephemeral(r)  (void *) (&r->uri_start)
 
+#define ngx_http_proxy_auth(r)  ((r)->method == NGX_HTTP_CONNECT)
```
It may be a good idea to re-indent these macros.

arut replied:
OK.
